hook!-full-game/addons/level_design_dock/plugin.gd

```diff
@@ -0,0 +1,111 @@
+tool
+extends EditorPlugin
+
+const INTERFACE_SCENE: = preload("res://addons/level_design_dock/LevelDesignDock.tscn")
+const LevelDesignDock: = preload("res://addons/level_design_dock/LevelDesignDock.gd")
+const PackedSceneButton: = preload("res://addons/level_design_dock/PackedSceneButton.gd")
+
+var _interface: LevelDesignDock
+var _current_instance: Node2D setget _set_current_instance
+var _current_parent: Node setget _set_current_parent
+var _current_scene: PackedScene setget _set_current_scene
+
+func _enter_tree() -> void:
+	_interface = INTERFACE_SCENE.instance()
+	add_control_to_dock(EditorPlugin.DOCK_SLOT_RIGHT_UR, _interface)
+	_interface.connect("packed_scene_button_created", self, "_on_Interface_PackedScene_Button_Created")
+
+
+func _exit_tree() -> void:
+	_interface.queue_free()
+
+
+func handles(object: Object) -> bool:
+	var can_handle = object is Node
+	if can_handle:
+		_set_current_parent(object)
+	return can_handle
+
+
+func forward_canvas_gui_input(event: InputEvent) -> bool:
+	var forward: = false
+	if event is InputEventMouseMotion:
+		_move_instance()
+		forward = true
+	elif event is InputEventMouseButton:
+		if event.button_index == BUTTON_LEFT:
+			forward = true
+			if event.is_pressed():
+				_create_instance()
+			else:
+				_clear_current_instance()
+	if event is InputEventKey:
+		if event.scancode == KEY_ESCAPE:
+			_clear_current_instance()
```

Comment:
If the idea behind these methods is to encapsulate ahead of time, I don't recommend it as you need to go look at each method to understand how the code works, while `_current_instance = null` would make sense

Reply:
The idea is to avoid having to check for implementations details by reading the method's name, any suggestions to make it clearer? I want to avoid direct changes in the variable as the future is unsure, `_current_instance = null` gives less space for better implementations further and seems like a candidate for refactoring later.

Reply:
Same as above, GDScript not being C++ or Java, you can add a setter anytime if need be. This is a private function, for a private variable, so it will only be in this script. You could replace it anytime.

_In general,_ I consider that refactoring is core to our work. I don't count the times where I've had to just rewrite stuff entirely because the design requirements changed unexpectedly. This happened with Power Sequencer for instance, but also for the Player in this game, and almost every project before it. That's why now I try to favor simpler code, working with properties directly [in Godot and Python] whenever possible.

If you use another text editor, it doesn't take time as long as the changes don't spread over a large codebase. In high level languages, you don't need to encapsulate or use patterns as much as in a lower-level language. Actually, a lot of design patterns are for C++, or for back-end, low-level software architecture. Most of the patterns from the gangs of 4 or Game Programming Patterns don't apply to gameplay programming with Godot. Even though it's interesting to learn them to better understand software architecture, and how things can work under the hood.